fix: Annotate PLR when when started status is reported

gbenhaim opened the pull request:
## 📝 Description of the Change

The use of the state label (which is mutable) for deciding when to report to the SCM that the PLR was started is flaky.

It was seen that the reconciler get events about PLRs with unexpected value for the state label. For example, after the status is reported to the SCM, and the label value is patched to "started", after serval reconcile iterations the label had the "queued" value again. This can happen because of unsafe patching done by controllers (not just the PAC controllers) which reconciles PLRs.

Introduce a new annotation for indicating the the status was reported to the SCM. By adding an annotation which is set once, we remove the risk that its value will get overwritten by other controllers (since maps are merged when patched, values are not getting removed unless explicitly defined in the patch
- https://datatracker.ietf.org/doc/html/rfc7386#section-2)

In addition, at the start of each reconcile loop, ensure that we operate on the latest version of the PLR and not using a stale value from the cache.

Set the state annotation and labels in pipelineascode only after the state was reported to the SCM. This is achieved by patching the PLR with the state label/annotation after it was created.

This change is needed in order to avoid a case where the watcher will report a status before PAC.

With this change the issue can't happen, since the watcher only reconciles PLRs that has the state annotation, so the first reconcile of the PLRs will happen only after PAC already sent the initial status to the SCM.

In addition, simplify the condition which checks if the "queued" state annotation/label should be set. We can only check if the PLR is pending, it doesn't matter if the reason is because of PAC's concurrency control or an external controller.

## 🔗 Linked GitHub Issue

Fixes #

## 👨🏻‍ Linked Jira

 https://issues.redhat.com/browse/SRVKP-8255

## 🚀 Type of Change

- [x] 🐛 Bug fix (`fix:`)
- [ ] ✨ New feature (`feat:`)
- [ ] 💥 Breaking change (`feat!:`, `fix!:`)
- [ ] 📚 Documentation update (`docs:`)
- [ ] ⚙️ Chore (`chore:`)
- [ ] 💅 Refactor (`refactor:`)
- [ ] 🔧 Enhancement (`enhance:`)

<!-- (update the title of the Pull Request accordingly) -->

## 🧪 Testing Strategy

- [ ] Unit tests
- [ ] Integration tests
- [ ] End-to-end tests
- [ ] Manual testing
- [ ] Not Applicable

## ✅ Submitter Checklist

- [ ] 📝 My commit messages are clear, informative, and follow the project's [How to write a git commit message guide](https://developers.google.com/blockly/guides/contribute/get-started/commits). **The [Gitlint](https://jorisroovers.com/gitlint/latest) linter ensures in CI it's properly validated**
- [ ] ✨ I have ensured my commit message prefix (e.g., `fix:`, `feat:`) matches the "Type of Change" I selected above.
- [ ] ♽ I have run `make test` and `make lint` locally to check for and fix any
      issues. For an efficient workflow, I have considered installing
      [pre-commit](https://pre-commit.com/) and running `pre-commit install` to
      automate these checks.
- [ ] 📖 I have added or updated documentation for any user-facing changes.
- [ ] 🧪 I have added sufficient unit tests for my code changes.
- [ ] 🎁 I have added end-to-end tests where feasible. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.
- [ ] 🔎 I have addressed any CI test flakiness or provided a clear reason to bypass it.
- [ ] If adding a provider feature, I have filled in the following and updated the provider documentation:
  - [ ] GitHub App
  - [ ] GitHub Webhook
  - [ ] Gitea/Forgejo
  - [ ] GitLab
  - [ ] Bitbucket Cloud
  - [ ] Bitbucket Data Center

gbenhaim committed:
fix: Annotate PLR when started status is reported

The use of the state label (which is mutable) for deciding when to report
to the SCM that the PLR was started is flaky.

It was seen that the reconciler get events about PLRs with unexpected
value for the state label. For example, after the status is reported
to the SCM, and the label value is patched to "started", after serval
reconcile iterations the label had the "queued" value again. This can
happen because of unsafe patching done by controllers
(not just the PAC controllers) which reconciles PLRs.

Introduce a new annotation for indicating the the status was reported
to the SCM. By adding an annotation which is set once, we remove the
risk that its value will get overwritten by other controllers
(since maps are merged when patched, values are not getting removed
unless explicitly defined in the patch
- https://datatracker.ietf.org/doc/html/rfc7386#section-2)

In addition, at the start of each reconcile loop, ensure that we
operate on the latest version of the PLR and not using a stale value
from the cache.

Assisted-By: Cursor
Signed-off-by: Gal Ben Haim <gbenhaim@redhat.com>

gbenhaim commented:
@zakisk can you please guide me on where should I add tests for this change?

zakisk commented on pkg/apis/pipelinesascode/keys/keys.go:
```diff
@@ -61,10 +61,11 @@ const (
+	GithubApplicationID    = "github-application-id"
+	GithubPrivateKey       = "github-private-key"
+	ResultsRecordSummary   = "results.tekton.dev/recordSummaryAnnotations"
+	SCMReportingPLRStarted = pipelinesascode.GroupName + "/scm-reporting-plr-started"
```
@gbenhaim shouldn't this be on line no 63 above? grouped with other annotation keys.

gbenhaim replied:
done

zakisk commented:
/ok-to-test

zakisk commented on pkg/reconciler/reconciler.go:
```diff
@@ -54,6 +54,23 @@ var (
 	ctx = info.StoreNS(ctx, system.Namespace())
 	logger := logging.FromContext(ctx).With("namespace", pr.GetNamespace())
+
+	logger.Infof("=== RECONCILE START: PipelineRun %s/%s (resourceVersion: %s) ===", pr.GetNamespace(), pr.GetName(), pr.GetResourceVersion())
```
can you please rephrase log messages here and below to align with log convention used across project?

gbenhaim replied:
done

gbenhaim commented on pkg/reconciler/reconciler.go:
```diff
@@ -69,13 +86,27 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, pr *tektonv1.PipelineRun
 		return r.updatePipelineRunToInProgress(ctx, logger, repo, pr)
+	} else {
+		logger.Infof("RECONCILE: Condition NOT met for %s/%s - reason='%s', state='%s' (startReported: %v)", pr.GetNamespace(), pr.GetName(), reason, state, startReported)
 	}
```
done

gbenhaim commented on pkg/reconciler/reconciler.go:
```diff
@@ -69,13 +86,27 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, pr *tektonv1.PipelineRun
+	scmReportingPLRStarted, exist := pr.GetAnnotations()[keys.SCMReportingPLRStarted]
+	if exist && scmReportingPLRStarted == "true" {
+		startReported = true
+	} else if !exist {
```
done

enarha commented:
/ok-to-test

gbenhaim commented:
/hold
Current implementation still has a race condition in it.
A situation can happen where the watcher reports the status first and only then pac reports its status. 
I think that the creation of the status should be moved to be handled by the watcher, except for the case where the pipelinerun creation failed, in this case the watcher has nothing to reconcile and the pac should report the error.

@zakisk @enarha WDYT?

zakisk commented:
> I think that the creation of the status should be moved to be handled by the watcher

but in normal cases (without Kueue) if reconciliation fails e.g. because of race condition we're facing and other cases which we might not know at this moment, we could end up not having git status at all. and  current setup is working fine so I think we should keep it as it is.

cc: @chmouel

zakisk commented:
> A situation can happen where the watcher reports the status first and only then pac reports its status.

I didn't understand how this can be happen?

gbenhaim commented:
> I didn't understand how this can be happen?
after pac create the plr nothing stops the watcher from reporting the status to github

gbenhaim commented:
> > I think that the creation of the status should be moved to be handled by the watcher
> 
> but in normal cases (without Kueue) if reconciliation fails e.g. because of race condition we're facing and other cases which we might not know at this moment, we could end up not having git status at all. and current setup is working fine so I think we should keep it as it is.
> 
> cc: @chmouel

what do you mean if reconciliation fails we won't have git status at all? the reporting from PAC can fail as well, I don't see how is it different.

zakisk commented:
> what do you mean if reconciliation fails we won't have git status at all? the reporting from PAC can fail as well, I don't see how is it different.

I am not saying that reporting would fail but if many controllers are patching PipelineRun, there will be many reconciliation loops and I don't know if they cause any error to reporting. (we also have PaC concurrency stuff as well) at this moment I don't know what would break but this will be big design change I guess.

zakisk commented:
@gbenhaim for unauthorized users, we create mirror PR to run E2E tests but when I am doing it I am getting gitlint error about your commit message, can you please adjust length of lines in your commit message? (every line should be less than 80 chars)

```
Gitlint..................................................................Failed
- hook id: gitlint
- exit code: 2

9: B1 Line exceeds max length (82>80): "the label had the "queued" value again. This can happen because of unsafe patching"
12: B1 Line exceeds max length (81>80): "Introduce a new annotation for indicating the the status was reported to the SCM."
13: B1 Line exceeds max length (81>80): "By adding an annotation which is set once, we remove the risk that its value will"
18: B1 Line exceeds max length (86>80): "In addition, at the start of each reconcile loop, ensure that we operate on the latest"
make: *** [Makefile:119: gitlint] Error 4

```

gbenhaim commented:
@zakisk can you maybe make me an authorized user :) ?

gbenhaim commented:
@zakisk I fixed the commit message

zakisk commented:
> @zakisk can you maybe make me an authorized user :) ?

I am just following the etiquette that we generally add people after some contributions 🙂

zakisk commented:
🚀 **Mirrored PR Created for E2E Testing**<br><br>A mirrored PR has been opened for end-to-end testing: [View PR](https://github.com/openshift-pipelines/pipelines-as-code/pull/2209)<br><br>⏳ Follow progress there for E2E results.<br>If you need to update the PR with new changes, please ask a maintainer to rerun `hack/mirror-pr.sh`.

gbenhaim committed:
fix: Set state only after status report

Set the state annotation and labels in pipelineascode only after
the state was reported to the SCM.

This is achieved by patching the PLR with the state label/annotation
after it was created.

This change is needed in order to avoid a case where the watcher will report
a status before PAC.

With this change the issue can't happen, since the watcher only reconciles
PLRs that has the state annotation, so the first reconcile of the PLRs
will happen only after PAC already sent the initial status to the SCM.

In addition, simplify the condition which checks if the "queued" state
annotation/label should be set. We can only check if the PLR is pending,
it doesn't matter if the reason is because of PAC's concurrency control
or an external controller.

Signed-off-by: Gal Ben Haim <gbenhaim@redhat.com>

gbenhaim commented:
@zakisk I pushed another commit that should solve the race condition I mentioned in earlier comments. It solves it without moving all the reporting to the SCM to the watcher.

zakisk commented on pkg/reconciler/reconciler.go:
```diff
@@ -54,6 +54,21 @@ var (
+		return fmt.Errorf("cannot get fresh PipelineRun: %w", err)
+	}
+
+	if freshPR.GetResourceVersion() != pr.GetResourceVersion() {
```
@gbenhaim there is similar check below on line no 125 in this file, if version isn't matched it should return.

gbenhaim replied:
I move it to the to of the function as me you and @enarha agreed today.

zakisk replied:
triggered CI!

gbenhaim committed:
fix: Skip reconciliation of stale PLRs

Instead of running the reconcile loop on the fresh pipeline, skip it
and let the next iterations to reconcile it.

Signed-off-by: Gal Ben Haim <gbenhaim@redhat.com>

enarha commented on pkg/pipelineascode/pipelineascode_test.go:
```diff
@@ -718,10 +718,19 @@ func TestRun(t *testing.T) {
 						secretName, ok := pr.GetAnnotations()[keys.GitAuthSecret]
 						assert.Assert(t, ok, "Cannot find secret %s on annotations", secretName)
 					}
-					if tt.concurrencyLimit > 0 || pr.Spec.Status == pipelinev1.PipelineRunSpecStatusPending {
```
Why remove the `tt.concurrencyLimit > 0` check? Assuming it was there for a reason, maybe add a separate `if` statement if you want to check the behavior without it.

gbenhaim replied:
When thr concurrency limit > 0, the pipeline run is pending, so checking both is redundent.

gbenhaim replied:
https://github.com/openshift-pipelines/pipelines-as-code/pull/2208/files#diff-7d95c98719a2b5fe21a2104db3b95cda3425ab76566525da1437521470481c79R180

enarha commented on pkg/pipelineascode/pipelineascode.go:
```diff
@@ -240,23 +236,31 @@ func (p *PacRun) startPR(ctx context.Context, match matcher.Match) (*tektonv1.Pi
+	} else {
+		// Mark that the start will be reported to the Git provider
+		patchAnnotations[keys.SCMReportingPLRStarted] = "true"
+		patchAnnotations[keys.State] = kubeinteraction.StateStarted
```
Why do we need to set the state here? I see it is otherwise the reconciler setting the state label and annotation. I understand why the SCMReportingPLRStarted was added here, but why change the existing behavior by also setting state here?

gbenhaim replied:
Because the plr isn't getting created with the state anymore, it should be set here as well to support plrs which are not pending.

The reconciler won't reconcile plrs without the state annotation https://github.com/openshift-pipelines/pipelines-as-code/blob/fa877bb9d07e25ca680664394aaada8eb1856c5d/pkg/reconciler/controller.go#L77

enarha replied:
OK, I see.

enarha reviewed (commented):
/lgtm

gbenhaim commented:
@zakisk what is the next step for getting this one merged?

gbenhaim commented:
> Caution
> 
> There are some errors in your PipelineRun template.
> 
> PipelineRun	Error
> git-clone	`error decoding yaml document: no kind "StepAction" is registered for version "tekton.dev/v1alpha1" in scheme "pkg/runtime/scheme.go:100"`

How this error is related to my change?

chmouel commented:
seems like it was an infra error.. not sure why the stepactions crds wasn't applied at the time you ran this...

chmouel commented:
/retest

chmouel commented:
I have done a code review and tried to "simulate" a random failure with concurrency in the code:

```diff
diff --git a/pkg/reconciler/reconciler.go b/pkg/reconciler/reconciler.go
index 58493f289..14dee5134 100644
--- a/pkg/reconciler/reconciler.go
+++ b/pkg/reconciler/reconciler.go
@@ -276,7 +276,30 @@ func (r *Reconciler) reportFinalStatus(ctx context.Context, logger *zap.SugaredL
 	return repo, nil
 }
 
+func randomError(prn string) error {
+	if strings.HasPrefix(prn, "test-gh-2") {
+		return fmt.Errorf("DEBUG: 😈 randomly failing this PipelineRun: %s", prn)
+	}
+	/* if we want to generate more randomness on load testing
+	   var b [8]byte
+	   if _, err := rand.Read(b[:]); err != nil {
+	     return fmt.Errorf("failed to generate random number: %w", err)
+	   }
+	   randomNumber := binary.LittleEndian.Uint64(b[:]) % 100
+
+	   if randomNumber < 50 {
+	    return fmt.Errorf("DEBUG: 😈 randomly failing this PipelineRun: %s", prn)
+	   }
+
+	*/
+	return nil
+}
+
 func (r *Reconciler) updatePipelineRunToInProgress(ctx context.Context, logger *zap.SugaredLogger, repo *v1alpha1.Repository, pr *tektonv1.PipelineRun) error {
+	if err := randomError(pr.GetName()); err != nil {
+		return err
+	}
+
 	pr, err := r.updatePipelineRunState(ctx, logger, pr, kubeinteraction.StateStarted)
 	if err != nil {
 		return fmt.Errorf("cannot update state: %w", err)
```

I thought it would fail to continue and get stuck forever when the concurrency get out of order because of [`d34206b` (#2208)](https://github.com/openshift-pipelines/pipelines-as-code/pull/2208/commits/d34206bed85d7bf4f6ab92091bb08f881f4a5f3b#diff-fc2bc6354d4506ced4adfd45dda2737dc917c217e282041d9af5bce57e514a85R68)

but it seems to work well and continue until the end, i haven't dug into the commit flow to understand how that happen tho but at least it works for me


https://github.com/user-attachments/assets/daa4b594-f5e7-4724-b0a1-4f3688cd4624

chmouel commented:
/lgtm

zakisk commented:
/lgtm
